421 Group Management

danhorst opened the pull request:
Fixes #421 

This PR lets you create groups and add users to groups in an administrative GUI.

_In this PR group membership doesn’t have any impact on other behavior in Hyku._

Granting membership to a `Hyku::Group` has not been integrated with `User#groups`, `Ability` or `Hyrax::Ability`. There are several questions that need to be resolved before that work can start:
- Are there “system groups”, like “Site Administrators” in `Hyku::Group`?
  - How tightly should the group attributes be tied to the `cancan` logic?
  - Are they generated by default?
  - Can they be deleted?
- Can groups be granted permission to act on a Work?
- Can groups be granted permission to act on a `Collection`?
- Can groups be granted permission to act on an `AdminSet`?

[See answers below](#issuecomment-272296038), discussion and implmentaion of other group features has moved to #610, #611, and #612.

---

### Initial group spec

@e61b42661c84857fc4b3bb8cd1009b8f94b171ba


### Get tests passing with a PORO group model

@8ab23859ce3fc7cbfaa8a003dc88f0b5480133d8


### Including ApplicationRecord as per Rails 5 convention

@c1183c4506f8616c448aa3f7af14ef3fd020c981


### Using ActiveRecord as a back-end for groups

@821ecef125b6bfbd5f1ddf4115b8be89494a7345


### Placeholder index page for group management

@f7960d52c39aa0d0f94b0843a8f247c46fecf369


### Simple table view for groups

@9ea9a19e862620ff85c955166149c386f703a034


### Basic new group form

@44fa2f481613fdf9731baf43ea27a1d76c19e0f8


### Update group behavior

@3d476edaeca26a935a9e92ca461dfe0b443e72f9


### Aligning from appearance with mock-ups

@cf5769dc8f8e0e311141d6ad136f12adaa7d18c9


### Ham-fisted title fix

@6500532b6dee65d2acd828f7337f0863c5d8eab5


### Vertical alignment adjustment as indicated on mock-up

@9ba4fa7d869a54d7a2c80d7d859fc917e9112eab


### Page for removing a group

@e8133a9fd06643a0aeccce5a79d33953ec30cedc


### Route, controller, and view for group membership

@52843fc5816c61162b07d8fd0a741ff92dae997a


### First pass at group tab navigation

@9b22baa74a2e29d4707e57d9be5fa428ac5b873e


### Refactoring tab construction with OpenStruct

@65d72571cd62d7e35e99d6ecd94423560982c721


### Change point of OpenStruct instantiation

@a0f862b0e85fa2e9d156285804ecc3d85b1b887c


### Disabling navigation controls for new groups

@063ccdd9f63f187ad8019e40e4d5621a65de7000


### Adding back-to-manage link

@861dc28ea87cfaac62dec57e7e400ab558a61f7d


### View refinements for group management

@205e2460c3e51b6a40d8ad9054e2133b0e320fd7


### Pagination for groups

@e27b9adaf17eea3ae48e247712d5772dc777202e


### Using Bootstrap 3 theme for pagination controls

@41e6cab32c378ce00e0198b0b0ec6d7fcad54c19

The Blacklight theme for kaminari wasn't preserving boostrap styles

### Per-page control for groups listing

@3d83dd37dee82a2a192171140d220418022c3e6c


### Adding search control to group index page

@51a7c398db274afdcc457355a5b71e9e89fee25f

Groups page looks the part

### Including additional fields in listing control forms

@0398d65e56fb441c530c61ea4fe20257a758af31

This way the forms don't stomp on each other.

### Adding group membership controls

@cdd388f6995206badbdeb834adbfd5eed0c410d9


### Adding filter user list controls

@96a176551a19108995c18a034d28845e191c809c


### Removing unset parameters from hidden fields

@7d066566941518566690d5370c1c93673d341c8c

This keeps the other parameters from being set to blank values when
the page size is changes or a filter query is applied

### Persisting group membership with rolify

@ca09abf771faebd42d3fde30b4d6dbd2ea75bc65


### Allowing group membership to be revoked from the UI

@08b6008b58f342818c0d3458d2c3a31d0af732fb


### Allow groups to be searched by keyword

@9a5f90489e8dc9e7771331cedeaf46c2b27d8931

Nothing fancy here, just simple SQL matching

### Simple control for resetting scope

@bae6b210f2ca943fdfe6f80f957c167300733f87


### Responsive tables FTW

@801aee761de2a66be11f7be9162f12ab931fa887


### Allow group members to be scoped by keyword

@a4f255e6fe623a383db7306cd2ec46163c1b26e0

Note that actual searching is restricted to the User model. If other
models can be added to a group then the :search_members method will need
to be extended.

### Make fewer assumptions about the behavior of .find

@fc57ef37fe9d655c79e07a236a928a8b1700e462

This change guarantees that #collect will always be called on an Array.

It also means that `member_class` has to do less work if it is
implementing an #find method (e.g. if it is not being provided by
ActiveRecord)

### JavaScript behavior for updating per-page selection

@e2f20ce37aae6cf5284813f936f3989d600f6d36

Using [BEM](http://getbem.com/) style class names. The 'js' prefix
indicates that a CSS class has attached JavaScript behavior.

The form controls are included in the HTML and then hidden by the
JavaScript in the spirit of progressive enhancement.

### Switch to a form for revoking group membership

@3f6a7571294dc0167de7f8c1899ccd04aa39a358

Using a form instead of a link with jQuery-UJS behaveior allows the
action to be performed without JavaScript. This also makes it simpler
to decorate the action with JavaScript behaviour.

Switching the route from member to collection also allows the remove
group API endpoint to accept batch changes.

### DRY-ing up admin controllers

@78af1a194f3cb03ada1bac7d5599ea1542d142fb


### Cleaning up flash usage and validations in groups controller

@5091df600ea601b780276fd71444e8fcba01e3a6


### JavaScript control for adding members to a group

@166e55a4d49d7cd2c4144c02539a16006d630aab


### Internationalization for groups

@18059312c59aac9fe35eed30c9a1126f5d49d35e

Also, fixing object marshalling for remove action.

### Resolving rubocop errors

@ffc2ca9bb2cb16949d57bd83b049d015ac0cd265


### Testing group search

@74e798053ca93f65da20c86884c419208b6f26b7


### Unit tests for Hyku::Group

@5f4421f4343d2ec152e43a541de8d0809a029f2e


### Filling in missing I18n string

@4a027e5dc064a2fc02949e7d73af27f37fefcce7


### Test syntax help from rubocop

@cfa2effed2cfe4a3cdd434f4c2b39447e5f9c220


### Testing user attributes in JSON expected by group AJAX functions

@2ad61543467994d4bc31150b9d3155d609214d0e


### Tests for Hyku::Group#number_of_users

@37c6441a89bf66dcba91df74ef63b90755c54e37


### Making test code slightly less verbose

@f259606c56b4c90658dc1b642c21071676b6269e

So, rubocop thinks my test module is too long. Maybe it's right.

### Stub for GroupsController tests

@922ecaf7b436b54690e545ca47c3ab7713b02b0e


### Tests for GroupsController

@aa305565e6641f29595558de1711f9136f2c9164

Fixed update action and simplified branching

### Testing GroupUsersController

@7f9dca763567d732c8617a537a6cf1ee9023213c


### View test for group index page

@7cc5fc419d6b0e1529b45e91ed1c3f335aefae7e


### I18n for tile of new group page

@bfb7cf488c16e57aa38335936b1be1c31e2ddb7f


### View test for new group page

@0fbb48bebceef083b1cbea65e7b07109f01b78c2


### Tests for edit group page

@3665abcb34b07ee521e305b22599d5d2187fa48c


### Spec for remove group page

@b03232d2617106115f6fe7edf556862e88953acb


### View tests for group members page

@c72594db02f36ec4db3423a7a4c4bcea719f7f5f

A lot of mocking was needed in order for params[:grop_id] to be set in
order for the path helper methods to resolve correctly.

### Appeasing rubocop

@d4718f98a18af9e877f25c55514ae88bcb42242e


### Increasing test coverage and simplifying mocking for group members view

@c983fba8c50e50b27e9a99a33915bb70e300d486


### Fixing regression failure in Admin::UsersController

@d22af1adebd138abe9b31f3013724742c1067c88

.deny_access is defined in Hyrax::UsersControllerBehavior and thus
overrides the .deny_access in AdminController. Re-defining it again
seems a bit silly but it works.

jcoyne commented:
>   Are there “system groups”, like “Site Administrators” in Hyku::Group?

I would think it would be best to have all groups in Hyku::Group.

>        How tightly should the group attributes be tied to the cancan logic?

I'd like for someone to be able to swap in a different group system (e.g. LDAP) without too much coding.

>        Are they generated by default?

I'm not sure what "they" is referencing here.

>        Can they be deleted?

Yes.

>    Can groups be granted permission to act on a Work?

Yes

>    Can groups be granted permission to act on a Collection?

Yes

>    Can groups be granted permission to act on an AdminSet?

Yes

danhorst commented on app/assets/stylesheets/hyku.scss:
```diff
@@ -124,3 +124,86 @@ footer.navbar {
+    width: 5em;
+  }
+
+  td:nth-child(3),
```
The convention in Hyrax is that the acceptable selector depth is 4. This issue has been resolved with changes to the hound config file.

danhorst commented on app/assets/stylesheets/hyku.scss:
```diff
@@ -124,3 +124,86 @@ footer.navbar {
+    width: 9em;
+  }
+
+  td:nth-child(4),
```
The convention in Hyrax is that the acceptable selector depth is 4. This issue has been resolved with changes to the hound config file.

danhorst commented on app/assets/stylesheets/hyku.scss:
```diff
@@ -124,3 +124,86 @@ footer.navbar {
+.list-scope {
+  position: relative;
+
+  .list-scope__query {
```
This is a valid BEM-style class name. The SCSS lint rule for `SelectorFormat` has been updated to `strict_BEM` to allow this.

danhorst commented on app/assets/stylesheets/hyku.scss:
```diff
@@ -124,3 +124,86 @@ footer.navbar {
+    padding-right: 1.8em;
+  }
+
+  .list-scope__reset {
```
This is a valid BEM-style class name. The SCSS lint rule for `SelectorFormat` has been updated to `strict_BEM` to allow this.

danhorst commented:
@jcoyne I moved further enhancements to groups into three tickets:
- #610 Make Groups Useful: Interoperability with other group providers
- #611 Make Groups Useful: Repository-level access controls
- #612 Make Groups Useful: Item-level access controls

jcoyne commented on config/locales/simple_form.en.yml:
```diff
@@ -2,6 +2,7 @@ en:
   simple_form:
     "yes": 'Yes'
     "no": 'No'
+    cancel: 'Cancel'
```
Can you add the corresponding `es` translations.  Using google translate would be good enough for a first pass

danhorst replied:
👍  See 88d87bb5a4356516bf20b5433813785aca184f36

jcoyne commented on config/locales/en.yml:
```diff
@@ -33,6 +33,58 @@ en:
     service:
       ok: 'OK'
       down: 'Down'
+  hyku:
```
Can you add the corresponding `es` translations.  Using google translate would be good enough for a first pass

jcoyne commented on app/views/kaminari/_prev_page.html.erb:
```diff
@@ -0,0 +1,3 @@
+<li>
+  <%= link_to_unless current_page.first?, raw(t 'views.pagination.previous'), url, :rel => 'prev', :remote => remote %>
```
Please use ruby 1.9 style hashes

jcoyne commented on app/views/kaminari/_next_page.html.erb:
```diff
@@ -0,0 +1,3 @@
+<li>
+  <%= link_to_unless current_page.last?, raw(t 'views.pagination.next'), url, :rel => 'next', :remote => remote %>
```
please use ruby 1.9 style hashes

jcoyne replied:
Why are you using `raw` here?

jcoyne commented on app/views/kaminari/_first_page.html.erb:
```diff
@@ -0,0 +1,3 @@
+<li>
+  <%= link_to_unless current_page.first?, raw(t 'views.pagination.first'), url, :remote => remote %>
```
Why raw? Please use ruby 1.9 style hashes.

jcoyne commented on app/views/kaminari/_gap.html.erb:
```diff
@@ -0,0 +1,3 @@
+<li class='disabled'>
+  <%= content_tag :a, raw(t 'views.pagination.truncate') %>
```
Why is this using raw?

jcoyne commented on app/views/kaminari/_last_page.html.erb:
```diff
@@ -0,0 +1,3 @@
+<li>
+  <%= link_to_unless current_page.last?, raw(t 'views.pagination.last'), url, {:remote => remote} %>
```
why `raw`?  Please use ruby 1.9 style hashes.  Please add spaces inside the braces.

jcoyne commented on config/locales/en.yml:
```diff
@@ -33,6 +33,58 @@ en:
+        describe_singular_quantity_html: "There is <b>1 group</b> in this repository."
+        action:
+          create: 'Create New Group'
+          edit_html: 'Edit group &amp; users'
```
If we make this just `&` and not `_html` will rails just autoescape it for us?

jcoyne commented on app/controllers/admin/groups_controller.rb:
```diff
@@ -0,0 +1,66 @@
+
+    def destroy
+      if @group.destroy
+        redirect_to admin_groups_path, notice: "#{@group.name} destroyed"
```
Add this flash to i19n.

jcoyne commented on app/controllers/admin/groups_controller.rb:
```diff
@@ -0,0 +1,66 @@
+        redirect_to admin_groups_path, notice: "#{@group.name} destroyed"
+      else
+        logger.error("Hyku::Group id:#{@group.id} could not be destroyed")
+        redirect_to admin_groups_path flash: { error: "#{@group.name} could not be destroyed" }
```
add this flash to i18n

jcoyne commented on app/controllers/admin/groups_controller.rb:
```diff
@@ -0,0 +1,66 @@
+
+    def update
+      if @group.update(group_params)
+        redirect_to admin_groups_path, notice: "#{@group.name} updated"
```
add flash to i18n

jcoyne commented on app/controllers/admin/groups_controller.rb:
```diff
@@ -0,0 +1,66 @@
+        redirect_to new_admin_group_path, alert: 'Groups must have a name'
+      else
+        logger.error('Valid Hyku::Group could not be created')
+        redirect_to new_admin_group_path, flash: { error: 'Group could not be created.' }
```
add flash to i18n

jcoyne commented on app/controllers/admin/groups_controller.rb:
```diff
@@ -0,0 +1,66 @@
+      if new_group.save
+        redirect_to admin_groups_path, notice: "#{new_group.name} created"
+      elsif new_group.invalid?
+        redirect_to new_admin_group_path, alert: 'Groups must have a name'
```
add flash to i18n

jcoyne commented on app/controllers/admin/groups_controller.rb:
```diff
@@ -0,0 +1,66 @@
+    def create
+      new_group = Hyku::Group.new(group_params)
+      if new_group.save
+        redirect_to admin_groups_path, notice: "#{new_group.name} created"
```
add flash to i18n

jcoyne commented on app/controllers/admin/users_controller.rb:
```diff
@@ -1,23 +1,15 @@
-
       def deny_access(_exception)
-        redirect_to main_app.root_url, alert: "You are not authorized to view this page"
+        redirect_to main_app.root_url, alert: 'You are not authorized to view this page'
```
add flash to i18n

jcoyne commented on app/controllers/admin_controller.rb:
```diff
@@ -0,0 +1,14 @@
+    end
+
+    def deny_access(_exception)
+      redirect_to main_app.root_url, alert: 'You are not authorized to view this page'
```
add flash to i18n

jcoyne commented on app/models/hyku/group.rb:
```diff
@@ -0,0 +1,45 @@
+module Hyku
+  class Group < ApplicationRecord
+    self.table_name = 'hyku_groups'
+    resourcify
```
What does this do? Can you add a comment?

danhorst commented on app/views/admin/groups/new.html.erb:
```diff
@@ -0,0 +1,20 @@
+<div class="panel-tab-wrapper">
+  <ul class="nav nav-tabs">
+    <li role="presentation" class="active">
+      <a href="javascript:void(0);"><%= t('hyku.admin.groups.nav.attributes') %></a>
```
In `app/views/admin/groups/new.html.erb` the tab anchors are disabled because you can neither add members to nor remove a group that doesn’t exist yet.

jcoyne replied:
Okay, that sounds fair.

danhorst commented on app/models/user.rb:
```diff
@@ -50,6 +50,13 @@ def groups
     []
   end
 
+  def as_json(*)
```
I added this model method because in `Hyrax::UsersControllerBehavior` it calls `.to_json` on a collection of `User` objects. ([source](https://github.com/projecthydra-labs/hyrax/blob/master/app/controllers/concerns/hyrax/users_controller_behavior.rb#L17)). I wanted a way to change the JSON output without overriding anything in Hyrax from Hyku.

Should Hyrax be changed to use JBuilder instead?

jcoyne replied:
@danhorst yes, hyrax should be changed to use JBuilder, but AFAICT that code in Hyrax is a collection of `SolrDocument`s not of `User` objects.

danhorst replied:
I'll look into it.

jcoyne commented:
@danhorst Looks like the PR didn't pass rubocop. Can you check this out?

danhorst commented:
@jcoyne I addressed the rubocop violations, Hound CI recommendations, rebased master, bumped the Hyrax version to get the new JSON serialization in [Hyrax 307](https://github.com/projecthydra-labs/hyrax/pull/307), and fixed any failing tests.

There is something going on with Travis right now though.

jeremyf commented:
I believe the `faraday_middleware` may have had an update released since this branch started.

jeremyf commented:
@danhorst It appears the @45722b4 addresses the stumbling block for faraday_middleware; You could review the commit message for the steps I took. Inline are the highlights:

Troubleshooting:

Reviewing the following Travis build:

    https://travis-ci.org/projecthydra-labs/hyku/builds/194965063

Saw the following:

```console
Bundler cannot continue installing faraday_middleware (0.11.0).
The checksum for the downloaded `faraday_middleware-0.11.0.gem` does
not match the checksum given by the server. This means the contents of
the downloaded gem is different from what was uploaded to the server,
and could be a potential security issue.
```

Reviewed https://rubygems.org/gems/faraday_middleware and saw 0.11.0.1
as being released likely during the time this PR was being created.
